From cdb9f7eb4aefdd278b3f347ab37391fb04c570c7 Mon Sep 17 00:00:00 2001
From: Eric Kidd <eric.m.kidd@dartmouth.edu>
Date: Thu, 21 May 2009 08:57:45 -0400
Subject: [PATCH] stc: Optionally output minFound & maxFound from FindText

When calling FindText, our application selects the entire matching
string, and optionally replaces it with another string.  Both of the
operations require us to know the start and end positions of the string
we've found.

This patch adds to optional output parameters to FindText: minFound and
maxFound.  If non-NULL, these parameters will be filled in with either
the minimum and maximum positions of the string we found, or
wxSTC_INVALID_POSITION if the underlying Scintilla call fails.

WARNING: This patch adds two new output parameters to FindText, which
is backwards-compatible for C++ applications, but may not be backwards-
compatible for all SWIG bindings.  I'd be happy to redraft this patch
to add a new FindTextRange function, or something along those lines, if
that would be more compatible.

---
 include/wx/stc/stc.h |    7 ++++++-
 src/stc/gen_iface.py |   25 ++++++++++++++++++++++---
 src/stc/stc.cpp      |   18 ++++++++++++++++--
 3 files changed, 44 insertions(+), 6 deletions(-)

diff --git a/include/wx/stc/stc.h b/include/wx/stc/stc.h
index 4d05112..0a6b322 100644
--- a/include/wx/stc/stc.h
+++ b/include/wx/stc/stc.h
@@ -2580,7 +2580,12 @@ public:
     int GetPrintColourMode() const;
 
     // Find some text in the document.
-    int FindText(int minPos, int maxPos, const wxString& text, int flags=0);
+    int FindText(int minPos, int maxPos, const wxString& text, int flags=0,
+#ifdef SWIG
+               int *OUTPUT, int *OUTPUT);
+#else
+               int *minFound = NULL, int *maxFound = NULL);
+#endif
 
     // On Windows, will draw the document into a display context such as a printer.
     int FormatRange(bool   doDraw,
diff --git a/src/stc/gen_iface.py b/src/stc/gen_iface.py
index dba5c5a..15eeeb1 100755
--- a/src/stc/gen_iface.py
+++ b/src/stc/gen_iface.py
@@ -354,18 +354,37 @@ methodOverrideMap = {
 
     'FindText' :
     (0,
-     '''int %s(int minPos, int maxPos, const wxString& text, int flags=0);''',
+     '''int %s(int minPos, int maxPos, const wxString& text, int flags=0,
+#ifdef SWIG
+               int *OUTPUT, int *OUTPUT);
+#else
+               int *minFound = NULL, int *maxFound = NULL);
+#endif''',
 
      '''int %s(int minPos, int maxPos,
                const wxString& text,
-               int flags) {
+               int flags,
+               int *minFound, int *maxFound) {
             TextToFind  ft;
             ft.chrg.cpMin = minPos;
             ft.chrg.cpMax = maxPos;
             wxWX2MBbuf buf = (wxWX2MBbuf)wx2stc(text);
             ft.lpstrText = (char*)(const char*)buf;
 
-            return SendMsg(%s, flags, (sptr_t)&ft);''',
+            int result = SendMsg(%s, flags, (sptr_t)&ft);
+            if (minFound) {
+                if (result == wxSTC_INVALID_POSITION)
+                    *minFound = wxSTC_INVALID_POSITION;
+                else
+                    *minFound = ft.chrgText.cpMin;
+            }
+            if (maxFound) {
+                if (result == wxSTC_INVALID_POSITION)
+                    *maxFound = wxSTC_INVALID_POSITION;
+                else
+                    *maxFound = ft.chrgText.cpMax;
+            }
+            return result;''',
      0),
 
     'FormatRange' :
diff --git a/src/stc/stc.cpp b/src/stc/stc.cpp
index 67189e7..f5ac665 100644
--- a/src/stc/stc.cpp
+++ b/src/stc/stc.cpp
@@ -1392,14 +1392,28 @@ int wxStyledTextCtrl::GetPrintColourMode() const
 // Find some text in the document.
 int wxStyledTextCtrl::FindText(int minPos, int maxPos,
                const wxString& text,
-               int flags) {
+               int flags,
+               int *minFound, int *maxFound) {
             TextToFind  ft;
             ft.chrg.cpMin = minPos;
             ft.chrg.cpMax = maxPos;
             wxWX2MBbuf buf = (wxWX2MBbuf)wx2stc(text);
             ft.lpstrText = (char*)(const char*)buf;
 
-            return SendMsg(2150, flags, (sptr_t)&ft);
+            int result = SendMsg(2150, flags, (sptr_t)&ft);
+            if (minFound) {
+                if (result == wxSTC_INVALID_POSITION)
+                    *minFound = wxSTC_INVALID_POSITION;
+                else
+                    *minFound = ft.chrgText.cpMin;
+            }
+            if (maxFound) {
+                if (result == wxSTC_INVALID_POSITION)
+                    *maxFound = wxSTC_INVALID_POSITION;
+                else
+                    *maxFound = ft.chrgText.cpMax;
+            }
+            return result;
 }
 
 // On Windows, will draw the document into a display context such as a printer.
-- 
1.6.1.2

